rfcs/0018-glides-trainsheet-events.md

```diff
@@ -0,0 +1,704 @@
+
+New restrictions on existing fields:
+- `type` (string): MUST be `"added"`.
+- `tripKey` ([TripKey](#TripKey)): will always be the added trip form, and never the scheduled trip form.
+- `startLocation` ([Location](#Location), conditionally required): where the trip will start. Required if `startTime` is specified.
+- `endLocation` ([Location](#Location), conditionally required): where the trip will end. Required if `endTime` is specified.
+- At least one of `startLocation` and `endLocation` is required.
+- At least one of `startTime`, `endTime`, and `previousTripKey` is required.
+- `dropped` SHOULD NOT be set.
+- All data SHOULD NOT be `"none"` or `"unset"`. (Instead, the fields would not be included).
+
+Glides SHOULD include the `cars` field to indicate the length of the train, even if no information is known about each car.
+
+### TripKey
+For scheduled trips, Glides does not have the same trip ID used by RTR. For added trips, while an ID is provided, it can be the case that an added trip from the Glides perspective is matched to a GTFS trip, in which case RTR will use the GTFS trip ID in publishing the information.
+
+In order to reduce event duplication, a TripKey is used to identify both added and scheduled trips.
+
+**Added Trip**
+- `serviceDate` (RFC3999 date): the service date for the trip.
+- `glidesId` (string): unique value representing this trip for the given `serviceDate`.
+```json
+{"serviceDate": "2023-01-20", "glidesId": "ADDED-123"}
+```
+**Scheduled Trip**
+- `serviceDate` (RFC3999 date): the service date for the trip.
+- `startLocation` ([Location](#Location)): where the trip is scheduled to start.
+- `endLocation` ([Location](#Location)): where the trip is scheduled to end.
+- `startTime` ([Time](#Time)): when the trip is scheduled to depart `startLocation`.
+- `endTime` ([Time](#Time)): when the trip is scheduled to arrive at `endLocation`.
+
+```json
+{
+  "serviceDate": "2023-01-20",
+  "startLocation": {"gtfsId": "place-mdftf"},
+  "endLocation": {"gtfsId": "place-heath"},
+  "startTime": "04:47:00",
+  "endTime": "05:34:00"
+}
+```
+*Notes*:
+- The added/scheduled distinction here is internal to Glides, and may not align with what is in other data sources. For example, an Added Trip in Glides may still correspond to a scheduled GTFS trip, and a Scheduled Trip may not appear in GTFS (due to track closures, for example).
+
+### TripUpdated
+This indicates that a trip was updated in some fashion:  consist, operators, departure time. The trip is either a scheduled trip, or an added trip that has already appeared in the events stream. Fields which are not present are not considered to be updated.
+
+Fields in the object:
+- `type` (string): `"updated"|"added"`. Determines whether this is a TripUpdated object or a  [TripAdded](#TripAdded) object. If it's `"added"`, then it's a [TripAdded](#TripAdded) object, see above. Subsequent updates to previously-added trips have `type` `"updated"`.
```

Comment:
Since this is the definition of TripUpdated, isn't it always "updated"? It's probably worth noting this in TripAdded as well.

Reply:
Because of the way I wrote TripAdded as extending TripUpdated, every TripAdded object is a valid TripUpdated object. That means that the "added" type needs to be valid here as well.

Benefits:
- The TripAdded can say "TripAdded objects are the same as TripUpdated, with the following [new restrictions]", instead of respecifying every field or describing `type` as a special case.
- The [Schema can do the same thing in a machine-readable way](https://github.com/mbta/schemas/pull/1/files#diff-be061f6ee95a4f66a4dd1c17600bf0e70221dd1b05b7558bf29bb9e0b62b7927R207).
Drawbacks:
- It's confusing
- It's harder to specify that updates with `type: added` have to follow additional restrictions instead of being simpler TripUpdated objects.

I'll look at whether I can clean it up.

Reply:
reworded in https://github.com/mbta/technology-docs/pull/18/commits/01331e10a7ae88a68d776f598c7eb05c79b91dac